conjure-java-core/src/main/java/com/palantir/conjure/java/services/dialogue/AsyncGenerator.java

```diff
@@ -276,51 +215,89 @@ public CodeBlock visitUnknown(String unknownType) {
-                        param.getArgName(),
+                        "$L.$L($S, plainSerDe.serialize$L($L));",
                         "_request",
-                        value.getParamId(),
-                        Objects.class);
+                        method,
+                        key,
+                        primitiveTypeName(primitiveType),
+                        argName);
             }
 
             @Override
-            public CodeBlock visitReference(com.palantir.conjure.spec.TypeName unused) {
-                // TODO(forozco): handle references correctly
-                return CodeBlock.of(
-                        "$L.putHeaderParams($S, $L.toString());", "_request", value.getParamId(), param.getArgName());
+            public CodeBlock visitOptional(OptionalType optionalType) {
+                return CodeBlock.builder()
+                        .add("$L.ifPresent(v -> {\n", argName)
+                        .add(generatePlainSerializer(method, key, "v", optionalType.getItemType()))
+                        .add("\n});")
+                        .build();
             }
 
             @Override
-            public CodeBlock visitList(ListType unused) {
-                return visitCollection();
+            public CodeBlock visitList(ListType value) {
+                return visitCollection(value.getItemType());
             }
 
             @Override
-            public CodeBlock visitSet(SetType unused) {
-                return visitCollection();
+            public CodeBlock visitSet(SetType value) {
+                return visitCollection(value.getItemType());
             }
 
-            private CodeBlock visitCollection() {
-                return CodeBlock.of(
-                        "$L.putAllHeaderParams($S, plainSerDe.serialize$L($L));",
-                        "_request",
-                        value.getParamId(),
-                        param.getType().accept(PlainSerializer.INSTANCE),
-                        param.getArgName());
+            @Override
+            public CodeBlock visitMap(MapType value) {
+                throw new SafeIllegalStateException("Maps can not be query parameters");
             }
 
             @Override
-            public CodeBlock visitDefault() {
-                return CodeBlock.of(
-                        "$L.putHeaderParams($S, plainSerDe.serialize$L($L));",
-                        "_request",
-                        value.getParamId(),
-                        param.getType().accept(PlainSerializer.INSTANCE),
-                        param.getArgName());
+            public CodeBlock visitReference(com.palantir.conjure.spec.TypeName typeName) {
+                TypeDefinition typeDef = typeNameResolver.apply(typeName);
+                if (typeDef.accept(TypeDefinitionVisitor.IS_ALIAS)) {
+                    return generatePlainSerializer(
+                            method,
+                            key,
+                            argName + ".get()",
```

Comment:
slight preference for this to not be a string type because it ends up containing snippets of java code when we recurse?